### PR TITLE
Improve state management.

### PR DESCRIPTION
**Pull request by di-wu**

### CODE REVIEW

File: metascore/main.mo [hunk 1]
```diff
@@ -102,17 +124,16 @@ shared ({caller = owner}) actor class Metascore() : async M.FullInterface {
             // Means that the caller was not the game canister itself.
             case (null) {};
             case (? g)  {
-                // TODO: update scores.
+                putGameRecord(caller, g.metadata, sortScores(scores));
```
[the remark this answers is not in the document]
Comment (jorgenbuilder) [reply metascore/main.mo:127]: I'm scared that sorting here will be very expensive, but I haven't done the math on it.

File: metascore/main.mo [hunk 2]
```diff
@@ -129,51 +150,123 @@ shared ({caller = owner}) actor class Metascore() : async M.FullInterface {
...
+    public query func getTop(n : Nat) : async [MPublic.Score] {
+        var top : [MPublic.Score] = [];
+        for ((p, scores) in scores.entries()) {
+            top := Array.append<MPublic.Score>(top, [(
+                p, GameRecord.totalScore(scores)
+            )]);
+        };
+        top;
+    };
+
```
Comment (jorgenbuilder) [inline metascore/main.mo:246]: This is a great start. Obviously we will need to working more on our query methods when we start building the frontend.

**Review by jorgenbuilder — APPROVED** [review]
On the whole this looks good. Feels like it could be tidied here and there, but nothing egregious.

The only thing I'm worried about here is the sort on `scoreUpdate`. I think some stress testing would be a good idea. I'd love to see at what point this setup breaks. I discussed some alternate data structure ideas, but the value sorted map feels like it could very well be good enough.

Another tool in the toolbox: we could play around with kicking the realtime score calculations to the frontend, label those calculations as "estimations" and then do periodic "certified" calculations on chain.

~~I'm going to merge this for now because it adds missing functionality, and we can follow up with some testing and refactoring.~~ Edit: since you marked this as draft I'll wait. My advice would be to merge this ASAP though. Looks good!

File: metascore/main.mo (hunk 1 above)
[the remark this answers is not in the document]
Comment (di-wu) [reply metascore/main.mo:147]: #32

File: metascore/Interface.mo [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        registerGame : shared MPublic.Metadata -> async ();
+        // @auth: admin
+        cron         : shared ()               -> async ();
+        addAdmin     : shared (Principal)      -> async ();
+        removeAdmin  : shared (Principal)      -> async ();
+        isAdmin      : query  (Principal)      -> async Bool;
+
+        // CHORE: add functions whenever it is public.
+    };
+};
```
Comment (jorgenbuilder) [inline metascore/Interface.mo:36]: 👀

**Review by jorgenbuilder — APPROVED** [review]

**Pull request merged**